subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/TimeTrackingProcessor.java

```diff
@@ -0,0 +1,141 @@
+            }
+        });
+    }
+
+    @VisibleForTesting
+    protected TimeTrackingProcessor(Processor delegate, AnnotationProcessingResult result, Factory<Timer> timeProvider) {
+        super(delegate);
+        this.processorClassName = determineClassName(delegate);
+        this.executionTimeByProcessor = result.getExecutionTimeByProcessor();
+        if (!executionTimeByProcessor.containsKey(processorClassName)) {
+            executionTimeByProcessor.put(processorClassName, 0L);
+        }
+        this.timeProvider = timeProvider;
+    }
+
+    private String determineClassName(Processor processor) {
+        if (processor instanceof DelegatingProcessor) {
+            return determineClassName(((DelegatingProcessor) processor).getDelegate());
+        }
+        return processor.getClass().getName();
+    }
+
+    @Override
+    public Set<String> getSupportedOptions() {
+        return track(new Factory<Set<String>>() {
+            @Override
+            public Set<String> create() {
+                return TimeTrackingProcessor.super.getSupportedOptions();
+            }
+        });
+    }
+
+    @Override
+    public Set<String> getSupportedAnnotationTypes() {
+        return track(new Factory<Set<String>>() {
+            @Override
+            public Set<String> create() {
+                return TimeTrackingProcessor.super.getSupportedAnnotationTypes();
+            }
+        });
+    }
+
+    @Override
+    public SourceVersion getSupportedSourceVersion() {
+        return track(new Factory<SourceVersion>() {
+            @Override
+            public SourceVersion create() {
+                return TimeTrackingProcessor.super.getSupportedSourceVersion();
+            }
+        });
+    }
+
+    @Override
+    public void init(final ProcessingEnvironment processingEnv) {
+        track(new Factory<Void>() {
+            @Override
+            public Void create() {
+                TimeTrackingProcessor.super.init(processingEnv);
+                return null;
+            }
+        });
+    }
+
+    @Override
+    public boolean process(final Set<? extends TypeElement> annotations, final RoundEnvironment roundEnv) {
+        return track(new Factory<Boolean>() {
+            @Override
+            public Boolean create() {
+                return TimeTrackingProcessor.super.process(annotations, roundEnv);
+            }
+        });
+    }
+
+    @Override
+    public Iterable<? extends Completion> getCompletions(final Element element, final AnnotationMirror annotation, final ExecutableElement member, final String userText) {
+        return track(new Factory<Iterable<? extends Completion>>() {
+            @Override
+            public Iterable<? extends Completion> create() {
+                return TimeTrackingProcessor.super.getCompletions(element, annotation, member, userText);
+            }
+        });
+    }
+
+    private <T> T track(Factory<T> factory) {
+        Timer timer = timeProvider.create();
```

Comment:
This is probably good enough, but if we wanted to be really precise we should accumulate nanoseconds and only convert to millis at the very end. I.e. our `Timer` seems to be missing a `stop` method.

Reply:
Instead of reinventing the wheel, I've switched to Guava's `Stopwatch` in ba7615153f1.